### PR TITLE
WT-4457 Add a maximum of dirty system buffers for the logging subsystem

### PR DESCRIPTION
**Pull request by ddanderson**

### CODE REVIEW

**Comment by ddanderson** [comment]
@sueloverso, can you please review?

File: src/log/log.c [hunk 1]
```diff
@@ -2090,6 +2092,18 @@ __wt_log_release(WT_SESSION_IMPL *session, WT_LOGSLOT *slot, bool *freep)
 	if (F_ISSET(slot, WT_SLOT_CLOSEFH))
 		__wt_cond_signal(session, conn->log_file_cond);
 
+	if (F_ISSET(slot, WT_SLOT_SYNC_DIRTY) && !F_ISSET(slot, WT_SLOT_SYNC) &&
+	    conn->log_dirty_max != 0) {
```
Comment (sueloverso) [inline src/log/log.c]: I think checking the `WT_SLOT_SYNC_DIRTY` flag is sufficient and we don't also need to check `conn->log_dirty_max`. That cannot be set with a 0 `log_dirty_max` (and if it changes via reconfigure, that's okay if this one falls on the side of syncing).

File: dist/api_data.py [hunk 2]
```diff
@@ -676,6 +676,11 @@ def __cmp__(self, other):
     Config('archive', 'true', r'''
         automatically archive unneeded log files''',
         type='boolean'),
+    Config('dirty_max', '0', r'''
+        maximum dirty system buffer cache usage, in bytes.  If non-zero,
+        schedule writes for dirty blocks belonging to the log in the
+        system buffer cache after that many bytes from the log are
+        written into the buffer cache.'''),
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply dist/api_data.py]: I like using `os_cache_dirty_max` as a percentage of the `file_max`, and have just pushed that change.  Fixed the other items you mention as well.

File: src/conn/conn_log.c [hunk 3]
```diff
@@ -267,6 +267,10 @@ __logmgr_config(
 	if (cval.val != 0)
 		FLD_SET(conn->log_flags, WT_CONN_LOG_ARCHIVE);
 
+	WT_RET(__wt_config_gets(session, cfg, "log.dirty_max", &cval));
+	if (cval.val != 0)
+		conn->log_dirty_max = (wt_off_t)cval.val;
+
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply src/conn/conn_log.c]: Yes, fixed.

File: src/log/log.c [hunk 4]
```diff
@@ -2090,6 +2092,18 @@ __wt_log_release(WT_SESSION_IMPL *session, WT_LOGSLOT *slot, bool *freep)
...
+		if ((ret = __wt_fsync(session, log->log_fh, false)) != 0) {
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply src/log/log.c]: Fixed.

File: src/log/log_slot.c [hunk 5]
```diff
@@ -203,6 +203,33 @@ __log_slot_close(
...
+		return;
+
+	conn = S2C(session);
+	log = conn->log;
+	current = &slot->slot_release_lsn;
+
+	if (current->l.file == log->dirty_lsn.l.file &&
+	    current->l.offset > log->dirty_lsn.l.offset &&
+	    current->l.offset - log->dirty_lsn.l.offset > conn->log_dirty_max) {
+		F_SET(slot, WT_SLOT_SYNC_DIRTY);
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply src/log/log_slot.c]: Comment added.

File: src/log/log.c [hunk 6]
```diff
@@ -2053,7 +2054,8 @@ __wt_log_release(WT_SESSION_IMPL *session, WT_LOGSLOT *slot, bool *freep)
 	 * responsible for freeing the slot in that case.  Otherwise the
 	 * worker thread will free it.
 	 */
-	if (!F_ISSET(slot, WT_SLOT_FLUSH | WT_SLOT_SYNC | WT_SLOT_SYNC_DIR)) {
+	if (!F_ISSET(slot, WT_SLOT_FLUSH | WT_SLOT_SYNC | WT_SLOT_SYNC_DIR |
+	    WT_SLOT_SYNC_DIRTY)) {
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply src/log/log.c]: Fixed.

**Comment by ddanderson** [comment]
@sueloverso, I think I've resolved the issues. Do you want to take another look?

**Comment by sulabhM** [comment]
@agorrod , the change LGTM, waiting on your opinion about the configure option naming.

File: dist/api_data.py [hunk 7]
```diff
@@ -676,6 +676,13 @@ def __cmp__(self, other):
     Config('archive', 'true', r'''
         automatically archive unneeded log files''',
         type='boolean'),
+    Config('os_cache_dirty_max', '0', r'''
```
Comment (agorrod) [inline dist/api_data.py]: I think `os_cache_dirty_pct` is better.

**Comment by agorrod** [comment]
Thanks - I'm happy with this apart from the configuration option naming change.

**Pull request merged**